Add a D implementation for std.math.ilogb

JohanEngelen opened the pull request:
The reasons for this PR:
- Gets rid of assembly code that is DMD-specific and does not work for LDC. Special case for Windows CRT is no longer needed.
- Avoiding the x87 FPU means this implementation is faster than the assembly code.
- Not calling C-runtime: ilogb can be inlined and the compiler can reason about it.
- Templated so it runs faster for floats and doubles

Some quick benchmarking showed that it is indeed quite a bit faster, especially (not surprising) for float and double. But the first reason is my main motivation.

I do not know how to do the casts to enable CTFE.
I added a few extra unittests to increase code coverage.

(edit: removed wrong comment about ARM)

Orvid commented on std/math.d:
```diff
@@ -2519,6 +2519,21 @@ unittest
 }
 
+private import core.bitop : bsr;
+static if (2*size_t.sizeof == ulong.sizeof)
```
Are you really sure this is actually needed?

JohanEngelen commented:
@Orvid The code does need a bsr that works for 64-bit numbers on all platforms (32 and 64 bit), which druntime does not provide. But you are right: it is not needed outside ilogb. To fix further function name resolution problems, I've renamed the 64-bit bsr to bsr_ulong. 

I've rebased the changes into one commit.

Safety0ff commented on std/math.d:
```diff
@@ -2532,74 +2532,162 @@ unittest
-            ret                         ;
+            size_t* uix = cast(size_t*)&x;
+            size_t msb = uix[MANTISSA_MSB];
+            size_t lsb = uix[MANTISSA_LSB];
```
Nitpick: I would prefer
```
uint msb = x >> 32;
if (msb)
    return bsr(msb) + 32;
else
    return bsr(cast(uint)x);
```
Over pointer casting.

JohanEngelen replied:
Thanks.
That makes it CTFE-able I think?

JohanEngelen replied:
@Safety0ff That also gets rid of Endianness-noise. I've amended the PR.

JohanEngelen commented:
The testbot failure is (as far as I can see) completely unrelated to this PR.

9il commented on std/math.d:
```diff
@@ -2532,74 +2532,161 @@ unittest
-    version (Win64_DMD_InlineAsm)
+    import core.bitop : bsr;
+    // Provide a bsr implementation for ulong in 32-bit mode
+    int bsr_ulong(ulong x) @trusted pure nothrow @nogc
```
Could you move this function outside from `ilogb`?
```D
static import core.bitop.bsr;
...
static if(size_t.sizeof == 4)
{
   int bsr_ulong(ulong x) @trusted pure nothrow @nogc
    {
         ...
    }
}
else
    private alias bsr_ulong = core.bitop.bsr;
```

The reason is to reduce template bloat.

ibuclaw replied:
There's also a ``bsr64`` implementation in ``std.numeric`` - it's private, however...  https://github.com/D-Programming-Language/phobos/blob/master/std/numeric.d#L85

9il replied:
... it is outdated. Looks like we need `package` `bsr64` in `std.math`.

JohanEngelen replied:
@ibuclaw How sure are you that the bsr64 in std.numeric works for both little endian and big endian? Using a union instead of bitshifts makes a large assumption about memory layout that I am not sure is justified.

ibuclaw commented on std/math.d:
```diff
@@ -2532,74 +2532,161 @@ unittest
+    static if (F.realFormat == RealFormat.ieeeExtended)
+    {
+        if (ex)
+        {   // If exponent is non-zero
```
I do believe the coding style of Phobos discourages comments nestled in on the same line as opening braces.  Please move this and others on the next line.

ibuclaw commented on std/math.d:
```diff
@@ -2532,74 +2532,161 @@ unittest
+        }
+        else if ((vl[MANTISSA_LSB]
+                       |(vl[MANTISSA_MSB] & 0x0000_FFFF_FFFF_FFFF)) == 0)
+        {   // vf is +-0.0
```
Strange indent

ibuclaw commented on std/math.d:
```diff
@@ -2532,74 +2532,161 @@ unittest
+    static if(is(Unqual!T == float))
+        int* vi = cast(int*)&vf;
+    else
+        long* vl = cast(long*)&vf;
```
Just an opinion, but a union would be much better than this list of casts.

JohanEngelen commented:
@ibuclaw and @9il Thanks for review, I've improved the changes, rebased, and are waiting for Travis results.

ibuclaw commented on std/math.d:
```diff
@@ -2532,74 +2548,154 @@ unittest
+        ushort[T.sizeof/2] vu;
+        uint[T.sizeof/4] vui;
+        static if(T.sizeof >= 8)
+            long[T.sizeof/8] vl;
```
Unfortunately, ``T.sizeof`` could be 12.  So a more complicated logical other than dividing by 8 is required.  What I may do in future is move this to be a part of floatTraits.

JohanEngelen replied:
In the case of `T.sizeof == 12` or 10, I was assuming this would truncate to long[1], which is fine? (the code knows about this)

JohanEngelen commented:
rebased for cleaner history.

JohanEngelen commented:
removed trailing whitespace, and rebased

9il commented:
Take a look at https://github.com/D-Programming-Language/phobos/pull/3499 .

JohanEngelen commented:
@9il  Do you mean that this PR should implement the same as #3499 ? That sounds backwards. I think this PR should be merged (it's been sitting here for 2+ months). Then when #3499 settles it can apply the same changes from frexp to ilogb.

9il commented:
No, only FYI. It can be merged with current implementation, thought.

JohanEngelen commented:
OK, sorry for my half-angry comment... :(
Thanks for the heads up. #3499 is certainly an improvement over the current hack (at least it concentrates the hack to one central point ;)

DmitryOlshansky commented:
@9il anything to review here or is it good to go?

9il commented:
LGTM except unittest.
@JohanEngelen Please generalise it with
```D
foreach(F; TypeTuple!(float, double real))
{
...
}
```

JohanEngelen commented:
@9il  I've tried to generalize the unittests, using a construct found elsewhere in math.d. (I am very new to D) Thanks for the patience.

9il commented:
Thank you for math PRs :)
Rebase all commits into one. @DmitryOlshansky probably will merge.

9il commented on std/math.d:
```diff
@@ -2664,34 +2760,50 @@ alias FP_ILOGBNAN = core.stdc.math.FP_ILOGBNAN;
+            tuple( -F.infinity, int.max     ),
+            tuple(  2.0       , 1           ),
+            tuple(  2.0001    , 1           ),
+            tuple(  1.9999    , 0           ),
```
To fix auto test failure use construction like `F(1.9999)` instead

9il replied:
other option:
```D
alias T = Tuple!(F, int);
T[] vals = 
[
 T(1.9999, 1),
....
];

```

9il commented:
@JohanEngelen You can use fixed size arrays to prevent GC allocations.

JohanEngelen commented:
@9il @DmitryOlshansky  All done now!

DmitryOlshansky commented:
Going to merge in day or so unless somebody rises an issue with it.
@ibuclaw ?

DmitryOlshansky commented:
Hm a potential regression with this is that it was callable wirg int by implicit conversion and now it won't match the constraint.

JohanEngelen commented:
@DmitryOlshansky  How is this solved in general in Phobos?

9il commented:
see also #3014 #3012

DmitryOlshansky commented:
In this case I'd add an overload for isIntegral that internally just does cast(double) for 32bit nunbers and below and cast(real) for everything else. 

On the other hand - log2 is trivially implementable for integers with test and bsr.

9il commented:
The overload should be mark with deprecated, thought.

DmitryOlshansky commented:
@9il I'm not so sure - it's quite customary to supply integer literal for FP function expecting implicit conversion. Otherwise we'd get welcome user with hard to understand 'doesn't match overload set'.

9il commented:
@DmitryOlshansky This is bad practice that can cause out-of-the-way errors in mathematical algorithms. However, if a function can have correct and optimal implementation for integer types (like `ilogb`), then an integer version of this function can be added to library without deprecation.

DmitryOlshansky commented:
@9il I know back in the university I've seen plenty of bugs in numerics due to use of int in C programs.

JohanEngelen commented:
Do you want me to add a unittest explicitly testing for a few (large) integer numbers?

DmitryOlshansky commented:
> Do you want me to add a unittest explicitly testing for a few (large) integer numbers?

Yes, integer version needs some tests to cover it.

ibuclaw commented on std/math.d:
```diff
@@ -2683,34 +2779,51 @@ alias FP_ILOGBNAN = core.stdc.math.FP_ILOGBNAN;
     assert(ilogb(nextUp(-float.min_normal)) == -127);
-    assert(ilogb(nextUp(-0.0F)) == -149);
-    static if (floatTraits!(real).realFormat == RealFormat.ieeeExtended) {
+    assert(ilogb(nextUp(-float(0.0))) == -149);
```
TIL

JohanEngelen commented:
@9il @DmitryOlshansky  I've added implementations and tests for integer ilogb arguments. What do you think?

DmitryOlshansky commented:
Lookin' good

DmitryOlshansky commented:
Auto-merge toggled on
